### PR TITLE
fix(ci): unblock pr-title workflow with inline regex check

### LINKED ISSUE #6751: fix(ci): pr-title workflow has never run — startup_failure on every PR since #6396
## Summary

The `Validate PR title` workflow added in #6396 (`feat(ci): force check pr title`, merged 2026-05-16) has **never run successfully**. Every invocation since merge returns `startup_failure` with 0 jobs created, so the check is silently inert — it does not appear on PR check suites and does not gate anything.

## Evidence

```
$ gh run list --repo zeroclaw-labs/zeroclaw --workflow pr-title.yml --limit 10
all 10 most recent runs: conclusion=startup_failure, jobs created: 0
```

Spot check on `JordanTheJet/fix-5155-delegate-prompt-mode` PR: only `Quality Gate` checks and `PR Path Labeler` appear; `Validate PR title` is absent from the check list even though Actions tab shows a run attempted.

## Root cause

This repo restricts third-party Actions via repo-level policy:

```
allowed_actions: "selected"
patterns_allowed:
  - actions/*
  - docker/*
  - dtolnay/rust-toolchain@*
  - Swatinem/rust-cache@*
  - softprops/action-gh-release@*
  - peaceiris/actions-gh-pages@*
```

`amannn/action-semantic-pull-request` is not in the allowlist, so GitHub refuses to start any job that uses it.

## Fix options

**A.** Add `amannn/action-semantic-pull-request@*` to the allowlist. Simplest, but expands the trusted-action surface on a workflow that uses `pull_request_target` (runs with write-capable token context).

**B.** Replace the action with an inline regex check on `${{ github.event.pull_request.title }}`. No external dependency, no allowlist change, no supply-chain risk on `pull_request_target`. ~10 lines of bash.

Recommending **B**. PR incoming.

## Related

- Introduced by: #6396
- Closes: this issue

### PR DESCRIPTION
**Pull request by JordanTheJet**
## Summary

- **Base branch:** `master`
- **What changed and why:**
    - what: replaced `amannn/action-semantic-pull-request` in `.github/workflows/pr-title.yml` with a local bash validator (`scripts/check-pr-title.sh`) plus a regression test harness (`scripts/check-pr-title.test.sh`) that runs ahead of the validator in CI.
    - why: the action is not on this repo's allowed-actions allowlist, so every run since #6396 merged has been `startup_failure` with 0 jobs — the check has never actually gated a PR (see #6751).
- **Scope boundary:** edits `.github/workflows/pr-title.yml` and adds two scripts under `scripts/`. No code, no other workflows.
- **Blast radius:** PR-title validation only. If the regex is wrong, PRs may be falsely rejected/accepted at title-edit time; no impact on builds, tests, or merges (this workflow is not in the required gate). The test harness (`scripts/check-pr-title.test.sh`) runs as the first CI step so a broken regex fails fast before it can mis-gate a PR.
- **Linked issue(s):** Closes #6751

## Validation Evidence

`scripts/check-pr-title.test.sh` covers 42 cases (accept + reject) including nested scopes, breaking-change indicator, unicode emoji, hyphens/dots/underscores in scopes, and a representative range of rejection cases (capitalized type, empty scope, missing colon, bang in wrong position). Local run: **42 passed, 0 failed.**

End-to-end verification will come from this PR's own `Validate PR title` run — the title `fix(ci): unblock pr-title workflow with inline regex check` is itself a positive test case.

## Security & Privacy Impact

- New permissions, capabilities, or file system access scope? `Yes (narrower)` — `permissions:` changed from `pull-requests: read` to `contents: read` (needed for `actions/checkout`; no longer needs PR write context).
- New external network calls? `No` — removes one third-party action; the remaining dependency is `actions/checkout` (first-party, SHA-pinned to `de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2`, matching every other workflow in the repo per FND-004 §5.4).
- Secrets / tokens / credentials handling changed? `Yes (reduced)` — `GITHUB_TOKEN` env removed from the step; the inline validator does not need it.
- PII, real identities, or personal data in diff, tests, fixtures, or docs? `No`

**Net positive:** trigger changed from `pull_request_target` → `pull_request`, so the workflow no longer runs in the base-repo context with elevated privileges. `PR_TITLE` is passed via env var (not interpolated into the `run:` shell text) and quoted inside the script, so titles with shell metacharacters are not an injection path.

## Compatibility

- Backward compatible? `Yes` — same workflow name, same triggers (sans `_target`), same semantics (`requireScope: true`-equivalent).
- Config / env / CLI surface changed? `No`

## Rollback

`git revert <squash sha>` (resolve post-merge) returns to the prior `amannn/action-semantic-pull-request` state. Since that workflow has never produced a real signal, revert risk is nil.

---

## Why a local validator over allowlisting the action

Considered: adding `amannn/action-semantic-pull-request@*` to the repo's selected-actions allowlist. Rejected because:

1. **Supply-chain risk under `pull_request_target`** — the original trigger ran in the base repo context with access to secrets. Switching to `pull_request` plus inline bash removes the entire third-party dependency for this workflow.
2. **The check is ~10 lines of bash.** Not worth an external dependency.
3. **Allowlist churn** — each new policy entry must be re-audited over time.

### CODE REVIEW

**Review by WareWolf-MoonWall — CHANGES_REQUESTED** [review]
# Review draft — #6752
# fix(ci): unblock pr-title workflow with inline regex check
# Proposed verdict: --request-changes
# Reviewer: WareWolf-MoonWall

---

This PR replaces `amannn/action-semantic-pull-request` (which has been causing
`startup_failure` on every run since #6396 because it is not on the repo's allowed-actions
allowlist) with an inline bash regex and a companion test script. It also switches the trigger
from `pull_request_target` to `pull_request` and removes the `GITHUB_TOKEN` env from the step.
I reviewed the workflow diff, the two new scripts, the action-pinning policy in FND-004 §5.4,
and the security implications of the trigger change. The approach is correct; one policy
violation needs to be fixed before this can land.

---

### 🔴 Blocking — `actions/checkout@v4` is a mutable tag, violating FND-004 §5.4

FND-004 §5.4 is explicit:

> **Policy:** All `uses:` references in workflow files must be pinned to a full commit SHA with a
> version comment. Mutable tags (`@v4`, `@main`, `@latest`) are not permitted. No exceptions.

The action being removed (`amannn/action-semantic-pull-request@48f256284bd46cdaab1048c3721360e808335d50`) was correctly SHA-pinned. The replacement step uses:

```
- uses: actions/checkout@v4
```

That is a mutable tag. Whatever `actions/checkout` pushes to the `v4` tag after this PR merges
is what this workflow will execute — without any review. The "No exceptions" language in FND-004
§5.4 exists precisely for first-party actions like `actions/checkout`, where the perceived trust
can make pinning feel unnecessary but the risk model is the same.

The fix is one line. Pin to the current `actions/checkout` v4 SHA:

```
- uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683  # v4.2.2
```

(Verify the current SHA from the `actions/checkout` releases page before applying; the SHA above
was the v4.2.2 release at the time of this review.)

---

### 🟢 What looks good — inline bash over allowlisting is the right security call

The rationale in the PR body is sound. `pull_request_target` runs in the base-repo context with
elevated permissions and access to secrets — pulling a third-party action there is substantially
riskier than for a standard `pull_request` trigger. The switch to `pull_request` plus inline bash
removes the entire supply-chain dependency for this workflow. A title validation check is ~10
lines of bash; an external action is not warranted, and the argument against allowlisting
(ongoing audit burden, each new entry needs re-review over time) is correct.

---

### 🟢 What looks good — `pull_request_target` → `pull_request` is a net security improvement

The original `pull_request_target` trigger was over-privileged for a workflow that only needed
to read a PR title. Switching to `pull_request` means this workflow runs in the fork context
without access to base-repo secrets. Combined with removing `GITHUB_TOKEN` from the step env
(the inline bash does not need it), the net permission surface is significantly smaller. The
`contents: read` permission is appropriate for a workflow that checks out the repo to run the
test and validation scripts.

---

### 🟢 What looks good — test script coverage is thorough

`check-pr-title.test.sh` covers 20 accept cases and 20 reject cases, including edge cases I
would expect to see in a real PR queue: deeply nested scopes (`feat(providers/compatible/openai)`),
breaking-change indicator (`fix(scope)!:`), unicode emoji in the description, scope with hyphens,
scope with dots, and a meaningful range of rejection cases (capitalized type, empty scope, space
in scope, bang in wrong position, space between type and scope, missing colon). The test harness
uses `assert_accept` / `assert_reject` helpers that capture failures and report them at the end
rather than failing fast — a good choice for a test that exercises many independent cases.
Running the test suite as a CI step before the actual title check means a broken regex is caught
before it gates the wrong PRs.

---

### 🟢 What looks good — `PR_TITLE` is passed via environment variable, not interpolated into shell

The workflow passes the PR title as:

```yaml
env:
  PR_TITLE: ${{ github.event.pull_request.title }}
run: ./scripts/check-pr-title.sh "$PR_TITLE"
```

`${{ github.event.pull_request.title }}` is set as an environment variable value — it is not
interpolated directly into the `run:` shell text — and `"$PR_TITLE"` is properly quoted in the
shell invocation. Inside `check-pr-title.sh`, `printf '%s' "$title"` quotes the variable
correctly before piping to `grep`. A PR title containing shell metacharacters does not create
an injection path.

---

### 🟡 Warning — rollback section has a blank commit SHA

The Rollback section reads: `git revert ` (backtick-enclosed SHA is empty — the template
placeholder was not filled). Since this workflow has never produced a real signal, the rollback
is trivially `git revert <squash sha>`, but the blank is a template artifact that should be
cleaned up before merge. Fill with `git revert <squash sha>` (resolve after squash-merge) or
note "resolve post-merge."

---

The approach here is correct and the security reasoning is sound. The single thing blocking
this is the `actions/checkout@v4` mutable tag — one line to fix, policy is explicit, and the
existing workflows already demonstrate the correct form. Pin the SHA and this should be clear
to approve.

**Verdict: request-changes.**

File: .github/workflows/pr-title.yml [hunk 1]
```diff
@@ -1,22 +1,24 @@
...
 permissions:
-  pull-requests: read
+  contents: read
 
 jobs:
   main:
     runs-on: ubuntu-latest
     steps:
-      - uses: amannn/action-semantic-pull-request@48f256284bd46cdaab1048c3721360e808335d50  # v6.1.1
+      - uses: actions/checkout@v4
```
Comment (Project516) [inline .github/workflows/pr-title.yml]: v6 is latest

**Review by Project516 — CHANGES_REQUESTED** [review]

**Review by Audacity88 — COMMENTED** [review]
@JordanTheJet I reviewed current head 470da94a9f9a72ac854d2b6df206b3f4f3342b76, #6751, the workflow/script diff, the existing review thread, the PR body, and the current checks. I agree with @WareWolf-MoonWall’s active blocker, so I do not want to duplicate it with another request-changes review from me, but I do want to confirm the shape of the remaining ask.

### 🟢 What looks good — replacing the blocked action is the right direction

The move away from `amannn/action-semantic-pull-request` makes sense. #6751 shows that the old workflow was failing at startup because the action was outside the selected-actions policy, so it was not giving the repository a real signal. A small local validator plus a regression test script is a much better fit for this check, and switching away from `pull_request_target` also reduces the permission surface for title validation.

### 🔴 Blocking — `actions/checkout` still needs a full SHA pin

The remaining issue is the one WareWolf already flagged: the new workflow uses `actions/checkout@v4`. As WareWolf’s review details, the project’s action-pinning policy requires `uses:` entries to be pinned to a full commit SHA with a version comment. The current `@v4` ref is still a mutable tag, so it needs the same treatment.

Project516’s inline note is also worth accounting for before choosing the exact pin: if we want the latest `actions/checkout` line, verify and pin that release’s full SHA; if we intentionally want to stay on v4, pin the v4 release SHA and keep the version comment. The important part is that the reviewed workflow names an immutable commit, not a moving tag.

### 🔵 Suggestion — refresh the PR body to match the current head

One small cleanup while touching the PR body: the current body still describes the older one-file version of the change. Please refresh the scope/security/compatibility text so it matches the current head: two scripts were added, the trigger changed to `pull_request`, permissions changed to `contents: read`, and `actions/checkout` remains an action dependency once pinned.

Once the checkout ref is pinned, the body is refreshed, and the branch is green, I expect this to be straightforward to approve.

**Comment by Audacity88** [comment]
@JordanTheJet Quick ping on this one: the only remaining blocker looks like pinning actions/checkout to a full SHA and refreshing the PR body for the current two-script shape. The broken title-check workflow is generating a lot of failed startup notices across branches, so I’d like to get this landed soon. Are you able to make that quick update? If not, I can take over with a small superseding PR.

**Comment by Audacity88** [comment]
@JordanTheJet Thanks for the quick update. I see current head 574f2c466a035b9113e497cd8b685fa6ff3c3008 now includes the checkout SHA pin and refreshed PR body.

@WareWolf-MoonWall since your request-changes review is the active maintainer block here, could you re-review when you have a chance? I am leaving that block for you rather than approving over it. Final merge readiness will still need the usual update/CI refresh because the branch is currently `BEHIND`.

**Review by WareWolf-MoonWall — APPROVED** [review]
Re-reviewing current head `574f2c466a035b9113e497cd8b685fa6ff3c3008`. My prior review was on `470da94a9f9a72ac854d2b6df206b3f4f3342b76`.

### ✅ Prior 🔴 block resolved — `actions/checkout` is now SHA-pinned

My blocking item was the mutable `actions/checkout@v4` tag, which violated FND-004 §5.4. The new diff uses:

```
- uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
```

SHA-pinned with a version comment. That satisfies FND-004 §5.4.

### ✅ Prior 🟢 assessments unchanged

The inline bash approach over allowlisting, `pull_request_target` → `pull_request` permission reduction, `PR_TITLE` passing via env rather than direct shell interpolation, and test suite coverage all remain as strong as I described in the first review. The scripts are unchanged.

### 🟡 Prior warning still outstanding — blank rollback SHA in PR body

The rollback section in the PR body (`git revert ` with an empty SHA placeholder) was not updated. This is a minor housekeeping item; I am not holding on it. Please fill in with "resolve post-merge" or the squash SHA after merge.

My hold is released. Approving.
